### PR TITLE
Rate limit refactor [DPP-1001]

### PR DESCRIPTION
**Pull request by simonmaxen-da**
A refactor of the rate limiting code base as follows:

A new ADT `LimitResult` that has two possible values `OverLimit` and `UnderLimit`

A new type `LimitResultCheck = (FullMethodName, IsStream) => LimitResult`

The `RateLimitingInterceptor` now takes a `List[LimitResultCheck]`

The following checks implement `LimitResultCheck` and contain supporting functions:

- `MemoryCheck`
- `ThreadpoolCheck`
- `StreamCheck`

Changes to `RateLimitingInterceptorSpec`  have been deliberately kept minimal to show the functioning is unchanged.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/LimitResult.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  def flatMap(f: Unit => LimitResult): LimitResult
+}
+
+object LimitResult {
+
+  type FullMethodName = String
+  type IsStream = Boolean
+  type LimitResultCheck = (FullMethodName, IsStream) => LimitResult
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/LimitResult.scala:22]: previously I would have used a trait for LimitResultCheck and then had the ability to give parameters meaningful names (like `fullMethodName` and `isStream`).  `FullMethodName` and `IsStream` are only there to document the types (as opposed to having `type LimitResultCheck = (String, Boolean) => LimitResult`.

[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/LimitResult.scala:22]: I understand the reasoning behind using the type alias, but I am questioning the type itself. 
This looks more extendable to me:
```
  type LimitResultCheck = MethodDescriptor[_, _] => LimitResult
```

I am just not sure `[_,_]` would actually work :)

[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/LimitResult.scala:22]: I think that would work, _and is nicer_.

[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/LimitResult.scala:22]: second argument wins the battle ;) Thanks!

**Review by skisel-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 2]
```diff
@@ -12,7 +12,13 @@ import com.daml.error.definitions.LedgerApiErrors.{
 }
 import com.daml.error.{ContextualizedErrorLogger, DamlContextualizedErrorLogger}
 import com.daml.metrics.{MetricName, Metrics}
-import com.daml.platform.apiserver.RateLimitingInterceptor._
+import com.daml.platform.apiserver.RateLimitingInterceptor.{
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:15]: Why do we still need this version of the RateLimitingInterceptor?

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/ratelimiting/LimitResultSpec.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  it should "compose under limit" in {
+    val actual: LimitResult = for {
+      _ <- underCheck()
+      _ <- underCheck()
+    } yield ()
+    actual shouldBe UnderLimit
+  }
+
+  it should "compose over limit" in {
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/ratelimiting/LimitResultSpec.scala:29]: We should also test that a subsequent underCheck does not cancel an overCheck

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 2 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:15]: This maybe a rebase recovery - removed.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/ratelimiting/LimitResultSpec.scala (hunk 3 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/ratelimiting/LimitResultSpec.scala:29]: Added to test.

**Review by mziolekda — APPROVED** [review]

**Pull request merged**